lib/manager/abstract.js

```diff
@@ -0,0 +1,228 @@
+
+const defaultSelector = async (devices) => {
+  return devices[0];
+};
+
+/**
+ * Abstract Device Manager
+ *
+ * @property {Logger} logger - logger with context
+ * @property {Network} network - bcoin network
+ * @property {Object} options - last fromOptions.
+ * @property {Boolean} opened - whether manager is open
+ * @property {AbstractDeviceWrapper}  selected - selected Device.
+ * @property {Set<string, AbstractDeviceManager>} cachedDevices
+ *                                                   - list of devices.
+ */
+
+class AbstractDeviceManager extends EventEmitter {
+  constructor(options) {
+    super();
+
+    this.logger = Logger.global.context('device-manager');
+    this.network = Network.primary;
+    this.options = {};
+
+    this.selector = defaultSelector;
+    this.opened = false;
+    this.selected = null;
+    this.cachedDevices = new Map();
+
+    if (options)
+      this.fromOptions(options);
+  }
+
+  /**
+   * Inject properties from options.
+   * @param {Object} options
+   * @param {Logger?} options.logger
+   * @param {Network?} options.network
+   */
+
+  fromOptions(options) {
+    assert(typeof options === 'object');
+
+    this.options = options;
+
+    if (options.selector != null) {
+      assert(typeof options.selector === 'function');
+      this.selector = options.selector;
+    }
+
+    if (options.logger != null) {
+      assert(typeof options.logger === 'object');
+      this.logger = options.logger.context('device-manager');
+    }
+
+    if (options.network != null)
+      this.network = Network.get(options.network);
+
+    return this;
+  }
+
+  /**
+   * Get vendor identifier.
+   * @returns {String}
+   */
+
+  get vendor() {
+    throw new Error('Abstract getter.');
+  }
+
+  /**
+   * Initialize manager.
+   * @returns {AbstractDeviceManager}
+   */
+
+  init() {
+    return this;
+  }
+
+  /**
+   * Listen to necessary events.
+   * @returns {AbstractDeviceManager}
+   */
+
+  bind() {
+    throw new Error('Abstract method.');
+  }
+
+  /**
+   * Remove listener.
+   * @returns {AbstractDeviceManager}
+   */
+
+  unbind() {
+    throw new Error('Abstract method.');
+  }
+
+  /**
+   * Open manager.
+   * @returns {Promise<AbstractDeviceManager>}
+   */
+
+  async open() {
+    throw new Error('Abstract method.');
+  }
+
+  /**
+   * Close manager.
+   * @returns {Promise<AbstractDeviceManager>}
+   */
+
+  async close() {
+    throw new Error('Abstract method.');
+  }
+
+  /**
+   * Select device.
+   * @param {DeviceWrapper?} device
+   * @returns {DeviceWrapper}
+   */
+
+  async selecetDevice(device) {
+    throw new Error('Abstract method.');
+  }
+
+  /**
+   * Deslect device.
+   * @returns {Boolean}
+   */
+
+  async deselectDevice() {
+    throw new Error('Abstract method.');
+  }
+
+  /**
+   * List available devices.
+   * @returns {DeviceWrapper[]}
+   */
+
+  getDevices() {
+    throw new Error('Abstract method.');
+  }
+
+  /*
+   * API
+   */
```

Comment:
Do the below API functions belong on this abstraction? `DeviceManager` does not inherit this class, and we do not export implementations of this class from the library.